### ISSUE TITLE
add to prometheus

### ISSUE THREAD

**Issue by Arnold1** [ben-st/go-mux#1]
hi, 
i currently play with your repo. i would like a counter in your example and expose it to prometheus/grafana. also how can you define an alarm (hipchat msg or email) in prometheus when the counter reaches 100 for example?

i changed your code as follows.
i send a request to http://localhost:8080/hello and watch http://localhost:8080/metrics i see the counter incrementing... but cannot see the counter in the prometheus ui on localhost:9090. any idea?
 
```
package main

import (
	"fmt"
	"math/rand"
	"net/http"
	"os"
	"time"

	"github.com/gorilla/handlers"
	"github.com/gorilla/mux"
	"github.com/prometheus/client_golang/prometheus"
	promhttp "github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/sirupsen/logrus"
)

var (
	// Counter vector to which we can attach labels. That creates many key-value
	// label combinations. So in our case we count requests by status code separetly.
	counter = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "hello_requests_total",
			Help: "Total number of /hello requests.",
		},
		[]string{"status"},
	)
)

func init() {
	prometheus.MustRegister(counter)
}

func main() {
	var log = logrus.New()
	// set the log output
	log.Out = os.Stdout
	// set the log level
	log.Level = logrus.DebugLevel
	// Routes consist of a path and a handler function.
	r := mux.NewRouter()
	// sample log.Info
	log.Info("http server is ready")
	// sample log.Debug
	log.Debug("i am only visible in debug mode\n")
	// exposes /hello endpoint with the helloHandler handler
	r.HandleFunc("/hello", helloHandler)
	// exposes /metrics endpoint with standard golang metrics used by prometheus
	r.Handle("/metrics", promhttp.Handler())
	// wrap a logger arount the mux router
	loggedRouter := handlers.LoggingHandler(os.Stdout, r)
	// wrap a prometheus instrument handler around the logger
	prometheusRouter := prometheus.InstrumentHandlerWithOpts(prometheus.SummaryOpts{}, loggedRouter)
	// Bind to a port and pass our loggedRouter in
	log.Fatal(http.ListenAndServe(":8080", prometheusRouter))

}

func doSomeWork() int {
	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)

	statusCodes := [...]int{
		http.StatusOK,
		http.StatusBadRequest,
		http.StatusUnauthorized,
		http.StatusInternalServerError,
	}
	return statusCodes[rand.Intn(len(statusCodes))]
}

func helloHandler(w http.ResponseWriter, r *http.Request) {
	var status int

	defer func(begun time.Time) {
		// hello_requests_total{status="200"} 2385
		counter.With(prometheus.Labels{
			"status": fmt.Sprint(status),
		}).Inc()
	}(time.Now())

	status = doSomeWork()
	w.WriteHeader(status)
	w.Write([]byte("Hello, World!\n"))
}
```

**Comment by utf18** [comment]
hey,

~~i guess the issue is that the http.ListenAndServe function is blocking and your doSomeWork() is never called.
i went around this by calling a goroutine which runs before the http server is started and is running concurrently.~~

i did not read it correctly. my bad i will have a look at your code and come back to you.
Maybe you have to register the counter in the function

i added some code ~~covering this~~, hope this helps.
https://github.com/ben-st/go-mux-example/commit/f0d5b681f2e9ae474e080d35b1bd9857fd4cb0de

<img width="567" alt="screen shot 2017-11-21 at 20 28 44" src="https://user-images.githubusercontent.com/18464684/33092898-ce345008-cefb-11e7-9024-656166304dec.png">

**Comment by Arnold1** [comment]
the endpoint http://localhost:8080/metrics shows me the following:
```
# HELP hello_requests_total Total number of /hello requests.
# TYPE hello_requests_total counter
hello_requests_total{status="200"} 1
hello_requests_total{status="400"} 2
hello_requests_total{status="500"} 1
```
but i cannot see any metrics like that in prometheus...do you know what i mean? do you see the same issue?

**Comment by Arnold1** [comment]
in the prometheus ui http://localhost:9090/graph i only see the following metrics to select:
```
go_
http_
net_
process_
prometheus_
scrape_
tsdb_
up
```

here is what i do:
```
$ CGO_ENABLED=0 GOOS=darwin go build -o mux
$ ./mux

$ docker run -p 9090:9090 -v `pwd`/prometheus/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus
```

**Comment by Arnold1** [comment]
here is what i do:
```
$ CGO_ENABLED=0 GOOS=darwin go build -o mux
$ ./mux

docker run -p 9090:9090 -v `pwd`/prometheus/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus
```

**Comment by utf18** [comment]
i think i know what's happening.
this setup works under the assumption that you run it with docker-compose. If you run it as you describe the "target" in the prometheus.yml will not exist.

this is due to the reason, that the docker-compose uses a docker network to make use of internal DNS, so the go-mux-example described in the compose file is actually a valid dns name inside of the compose network.

i don't think your setup will work, because docker is not intended to connect from inside the container to the outer host.

You could try with --net=host, but i would recommend that you either run mux in a container oder run the prometheus outside of the container to test your changes.

you can build a mux docker image with:
(assuming that you build it outside with GOOS=linux)

FROM golang:alpine
ADD .mux /mux
RUN chmod 0755 /mux
EXPORT 8080
CMD ["/mux"]

and then build it with 
docker build -t mux:v1 .

after that you can change my image in the compose file with mux:v1 and everything should work fine or start them both inside of a container on your own.


tl:dr if your metrics are working in the /metrics endpoint you will have to take a hard look on prometheus in the /targets.
should look like this:

<img width="1602" alt="screen shot 2017-11-22 at 08 31 58" src="https://user-images.githubusercontent.com/18464684/33115027-b04b401a-cf5f-11e7-8f40-5204dc814d14.png">

**Comment by Arnold1** [comment]
works now:

my docker file:
```
FROM golang:1.9.2

RUN apt-get update && apt-get install -y git

RUN mkdir -p /home/go-mux-example
WORKDIR /home/go-mux-example

COPY main.go /home/go-mux-example
COPY handlers.go /home/go-mux-example

RUN go get -v github.com/gorilla/handlers
RUN go get -v github.com/gorilla/mux
RUN go get -v github.com/prometheus/client_golang/prometheus
RUN go get -v github.com/sirupsen/logrus

RUN chmod 0755 /home/go-mux-example
RUN CGO_ENABLED=0 GOOS=linux go build -o mux

EXPOSE 8080
CMD ["/home/go-mux-example/mux"]
```
and then build the docker with
```docker build -t mux:v1 .```

prometheus.yml file:
```
# my global config
global:
    scrape_interval:     15s
    evaluation_interval: 15s

    external_labels:
        monitor: 'myapp-monitor'

rule_files:
    # - "first.rules"
    # - "second.rules"

scrape_configs:
    - job_name: 'prometheus'
      static_configs:
          - targets: ['localhost:9090']

    - job_name: 'mux'
      static_configs:
          - targets: ['go-mux-example:8080']
```

docker compose file:
```
version: "2"

services:

  prometheus-server:
    image: prom/prometheus
    ports:
      - 9090:9090
    volumes:
      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml

  grafana:
    image: grafana/grafana
    ports:
      - 3000:3000
    environment:
      - GF_SECURITY_ADMIN_PASSWORD=password
    links:
      - prometheus-server:prometheus

  go-mux-example:
    image: mux:v1
    ports:
      - 8080:8080
```
than start docker compose:
```docker-compose up```

here is what i see:
<img width="1292" alt="metrics" src="https://user-images.githubusercontent.com/3844226/33135177-a99227d8-cf6f-11e7-9886-5eff6775005a.png">

<img width="1382" alt="screen shot 2017-11-22 at 10 45 39 am" src="https://user-images.githubusercontent.com/3844226/33136085-5225f030-cf72-11e7-9a7f-5aa681f18b63.png">

<img width="1346" alt="docker_compose" src="https://user-images.githubusercontent.com/3844226/33135218-cde2b738-cf6f-11e7-8cd3-78c6f31f2c91.png">

ok i can see the inc_value and hello_requests_total in here: http://localhost:9090/graph

<img width="1383" alt="screen shot 2017-11-22 at 10 46 53 am" src="https://user-images.githubusercontent.com/3844226/33136141-7d718cd6-cf72-11e7-8863-171a33706f2a.png">

<img width="1384" alt="screen shot 2017-11-22 at 10 48 03 am" src="https://user-images.githubusercontent.com/3844226/33136207-aa959a0e-cf72-11e7-9780-fbb36aa94f08.png">

**Comment by Arnold1** [comment]
do you know how to add alerts to prometheus which sends out an email and hipchat msg? for example: if counter reaches 50 -> send alert

**Comment by utf18** [comment]
nice 👍 

for the alerting part:
you have to have the alertmanager up and running:
`https://prometheus.io/docs/alerting/configuration/#<email_config>`
`https://github.com/prometheus/alertmanager`
and for the rules alerting the events you want to trigger:
`https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/`

or else get yourself started with the maybe simpler alerting system from grafana itself and alert on charts.
`http://docs.grafana.org/alerting/notifications/`

if i have the time, i can add a basic email alerting setup with a local mailserver or webhook to demonstrate, but for now the alerting depends on what you want to do.

i would suggest you add the alertmanager to the compose file and play with it a little bit.

Feel free to close the issue, if this covers your questions.

all the best,

benjamin
